EIPS/eip-4804.md

```diff
@@ -0,0 +1,163 @@
+
+#### Auto Mode
+
+The auto mode is the default mode to resolve (also applies when the "resolveMode" method is unavailable in the target contract). In the auto mode, if **path** is empty, then the protocol will call the target contract with empty calldata. Otherwise, the calldata of the EVM message will use standard Solidity contract ABI, where
+
+- **method** is a string of function method be called
+- **argument_i** is the ith argument of the method. If **type** is specified, the value will be translated to the corresponding type. The protocol currently supports the basic types such as uint256, bytes32, address, and bytes. If **type** is not specified, then the type will be automatically detected using the following rule in a sequential way:
+
+1. **type** = "uint256", if **value** is numeric; or
+2. **type**="bytes32", if **value** is in the form of 0x+32-byte-data hex; or
+3. **type**="address", if **value** is in the form of 0x+20-byte-data hex; or
+4. **type**="address", if **value** is in the form of **name**.**nsProviderSuffix**. In this case, the actual value of the argument will be obtained from **nsProviderSuffix**, e.g., eth, w3q, etc.
+5. else **type**="bytes"
+
+Note that if **method** does not exist, i.e., **path** is empty or "/", then the contract will be called with empty calldata.
+
+## Rationale
+
+The purpose of the proposal is to add a decentralized presentation layer for Ethereum.  With the layer, we are able to render any web content (including HTML/CSS, JPG/PNG/SVG, etc) on-chain using human-readable URLs, and thus EVM can be served as decentralized Backend.  The design of the standard is based on the following principles:
+
+- **Human-readable**.  The Web3 URL should be easily recognized by human similar to Web2 URL (http://).  As a result, we support names from name services to replace address for better readability.  In addition, instead of using calldata in hex, we use human-readable method + arguments and translate them to calldata for better readability.
+
+- **Maximum-Compatible with HTTP-URL standard**.  The Web3 URL should be compatible with HTTP-URL standard including relative pathing, query, fragment, etc so that the support of existing HTTP-URL (e.g., by browser) can be easily extended to Web3 URL with minimal modification.  This also means that existing Web2 users can easily migrate to Web3 with minimal extra knowledge of this standard.
+
+- **Simple**.  Instead of providing explict types in arguments, we use a "maximum likelihood" principle of auto-detecting the types of the arguments such as address, bytes32, and uint256.  This could greatly minimize the length of URL, while avoiding confusion.  In addition, explict types are also supported to clear the confusion if necessary.
+
+- **Flexible**.  The contract is able to override the encoding rule so that the contract has fine-control of understanding the actual Web resources that the users want to locate.
+
+### Examples
```

Comment:
This whole section could actually go in the specification section, instead of rationale.

Reply:
Done.  Thanks for the comment.